### ISSUE TITLE
Compiling/validating shaders asynchronously

### ISSUE THREAD

**Issue by frzi** [gpuweb/gpuweb#6]
Many developers will agree that freezing the interface the user interacts with results in a bad user experience. It is for this reason that developers may choose to perform some tasks asynchronously. Currently with WebGL compiling a shader / linking a program can freeze the browser for a short while; depending on the complexity of the shader and the user's hardware. On mobile this is usually way more noticeable.

Some would argue it's up to the developer to perform all shader generating, compiling and program linking at the very start of their web app or game. But there are valid reasons to generate and compile shaders at a later moment. E.g. loading in a new model at a later time (streaming?), or the user changing the game's quality settings during runtime. Or perhaps the app performs computations based on the user's input.

This is why I propose that, whatever function will compile the shader, should be either done forcefully asynchronously, or at least provide the option to do so, to prevent the browser from temporarily freezing. Said function would return a Promise, passing the compiled and validated shader to the resolve function and passing any error messages to the reject function.

Promises are *The Future of Async Javascript™* and it only seemed logical to choose Promises over callback functions. With the introduction of async functions (already implemented in some browsers!) I believe Promises have become a more strong and important feature of the Javascript language.

Continuing Apple's example of using the Metal API, a common path to compile a shader (library) may look something like this (ignoring any error handling):
```js
async function createRenderPipelineState() {
    const source = await fetch(/* url */).then(response => response.text())
    const library = await gpu.createLibrary(source)
    const vertexF = library.functionWithName('vertex_main')
    const fragmentF = library.functionWithName('fragment_main')
    /** etc **/
    return gpu.createRenderPipelineState(pipelineDescriptor);
}
```

**Comment by othermaciej** [comment]
For the sake of completeness it's worth noting that multiple threads being able to access the GPU is an alternative to asynchronous interfaces, and that's the path most of the native modern graphics APIs take. Multithreading gives you some abilities that pure async can't as easily, such as ability to render multiple frames ahead simultaneously. So it's probably a feature we'll have to support. At that point, async interfaces become just a convenience and we'll have to evaluate whether they are worth it.

**Comment by grovesNL** [comment]
@othermaciej: Could you clarify what multithreading might look like in the above scenario? Do you mean through use of workers or some other means?

The native API implementation can use multithreading to access the GPU however it prefers. However the API calls cannot be synchronous without blocking the JavaScript thread. So `Promise`/callback-based interfaces are necessary to avoid blocking, because there is no simple way to achieve multithreading in the API caller unless each call is run on a separate worker.

For example, using promises:

```
Promise.all([ createShader(vertexShaderSource), createShader(fragmentShaderSource) ])
    .then(shaderHandles => createProgram(shaderHandles))
    .then(programHandle => /* render using created program, etc. */)`
```

Here `createShader` and `createProgram` each return a `Promise`. The first `then` creates a program after the shaders have been compiled. The second `then` is executed after the program has been created. [Here's a quick example in JSFiddle](https://jsfiddle.net/grovesNL/7btfbed4/).

The JavaScript thread is not blocked while the programs compile, and the native API implementation could choose to do work concurrently wherever possible. [Note that WebGL is beginning to support `Promise` return values through extensions](https://github.com/KhronosGroup/WebGL/pull/2079).

**Comment by othermaciej** [comment]
@grovesNL The multi-threading based approach would be that multiple threads (either using Workers or some new threading mechanism) cans hare the same WebGPU context. Thus, you can use a dedicated Worker to compile and upload all your shaders, textures and other resources. Or more than one if you want. That way all the uploading of GPU resources can be done without blocking the main UI thread, without the need for explicitly asynchronous API. In fact, one could even use Workers plus message passing to build an async API on top.

In most contexts, I strongly prefer async API to "do it synchronous on a worker", but in this case it's likely that the API will need to be multithreading-oriented anyway, since parallel scene setup is a big part of how modern graphics APIs get their performance advantage at the native level. So it might be worth seeing how that works out before making a decision about adding async APIs.

With WebGL and even the underlying OpenGL, a lot of stuff has to be single-threaded so the same solutions don't necessarily apply.

**Comment by grovesNL** [comment]
@othermaciej Agreed, workers being able to interact with the same context would be great. Historically it's been difficult to efficiently pass data back and forth between workers, so that could present some challenges in this area.

Mozilla has experimented with some related ideas: [`SharedArrayBuffer`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer) to share data between workers and [`OffscreenCanvas`](https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas) to render to a WebGL context within a worker.

**Comment by grorg** [comment]
FWIW, I like the idea of online shader compilation returning promises (or async/await from ES2017). Even though we'll probably need multithreaded/worker access, it would be nice if a simple single threaded application could avoid blocking the UI.

**Comment by Kangz** [comment]
Async shader / pipeline compilation would clearly be a big win, and it is something that has been discussed for WebGL before (although no concrete proposal was accepted).

I'd like to point out that in our (Google) prototype, all NXT objects are essentially promises but you can use them immediately, including shader modules and pipelines. If needed the NXT "server" could offload compilation to multiple threads.

At application load time, you would create a bunch of pipelines, get their underlying promise and wait for Promise.all to know NXT finished creating them in the background and that you can start actually rendering things without stalls on the NXT server side.

**Comment by devshgraphicsprogramming** [comment]
This whole issue was encountered by game developers in the last decade and could be side-stepped entirely if webGPU ingested intermediate representation such as SPIR-V.

That is one of the reason why OpenGL developers went crazy over Vulkan.

**Comment by kainino0x** [comment]
Regardless of input format, WebGPU will still need to do validation because it is a secure environment running completely untrusted application code, including untrusted shader code. And we will still have to translate it to the underlying driver's representation (HLSL, DXIL, MetalSL, Vulkan SPIR-V) and the driver will still have to translate to machine code.

**Comment by devshgraphicsprogramming** [comment]
it will be easier for you do validate and downcompile from SPIR-V only input, and you can contribute to SPIRV-Cross for cross-compilation.

Lastly you only need cross compilation to DXIL and MetalSL in that case, and that's only in the event that Vulkan performance and reliability on Microsoft systems is egregious.

In the general sense its easier to reason about and validate untrusted code in the AST form which is really easy to build from an IL.

**Comment by kainino0x** [comment]
I agree, though the group has been debating this topic for a long time. By
the way, we do use SPIRV-Cross in Dawn.

**Comment by devshgraphicsprogramming** [comment]
I'd imagine in the end you'd only need Spir-V + MetalSL since the concepts of queues, passes, memory objects and events do not map easily to OpenGL ES and < DX 11.

If you keep WebGPU implementable on last-gen APIs (ES, GL, <DX11) you will needlessly throttle its performance,featureset and/or overcompilate the implementation, in which case you will have not much to offer over webGL 2.0 except for possibly compute+SSBOs.

**Comment by litherum** [comment]
The discussion of whether or not we want asynchronous compiles/validations (we do) is completely orthogonal from which language is used to represent the shader. Let’s stay on-topic, please.

**Comment by devshgraphicsprogramming** [comment]
TL;DR a Browser implementation of webGPU with an OpenGL backend would need a fixed pool of OpenGL threads and contexts for any asynchronous work.

**Comment by Kangz** [comment]
WebGPU doesn't target OpenGL or D3D11. Even with a text-based language, we expect the cost of the lowering to GPU ISA and aggressive driver optimizations to far outweigh the translation cost.

**Comment by devshgraphicsprogramming** [comment]
So Vulkan, DX12 and Metal are only possible backend implementations?

> we expect the cost of the lowering to GPU ISA and aggressive driver optimizations to far outweigh the translation cost.

Driver won't actually do that many aggressive optimizations after ingesting SPIR-V or DXIL, bulk of the work is actually done on GLSL-> SPIR-V side

**Comment by kainino0x** [comment]
We have been discussing this topic for a very long time, and believe me, all of these points have been brought up. You need not worry that SPIR-V is underrepresented in the discussions in the group; Firefox and Chrome are actively supporting it.

I will point out that, because WebGPU has to do a lot more shader ingestion work than a driver would, and because of overhead in various systems inside the browser (JS-C++ bindings, interprocess communication, additional validation and bounds check generation), WebGPU's shader ingestion will not be as fast as drivers' shader ingestion regardless of language chosen.

AFAIK, everyone in the group agrees that shader/pipeline compilation should be asynchronous since it is known to be a relatively slow operation. Currently this takes the form of an opaque "future"-like object that represents the future compiled shader/pipeline (much like most objects in the API), though it could become a Promise.

**Comment by devshgraphicsprogramming** [comment]
I think you should create entire Graphics Pipelines asynchronously, not just shaders... driver does shader instancing+alteration if you use different MSAA counts in your framebuffers, vertex formats, etc.
https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#pipelines-graphics
Nvidia actually has a silly function for the RTX pipelines because they are so heavy (shader tables)
https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#vkCompileDeferredNV

The whole "shader only, is not sufficient for real-under-the-hood-linking" was the reason why in OpenGL you had to do dummy draws with all FBO and VAO combinations to make sure instances of your shaders were precompiled for all Fixed Function States.
(on NVIDIA, ARB_debug will actually give you nice performance warnings if you suddenly use a GLSL program with a new vertex format or FBO layout that a shader needed to be recompiled)

**Comment by kainino0x** [comment]
Yes, I alluded to that too:

> AFAIK, everyone in the group agrees that shader/pipeline compilation should be asynchronous

**Comment by kainino0x** [comment]
I'm pretty sure this is resolved: shader and pipeline creation is asynchronous. Please file a new, more specific bug if anything is unresolved here.
